### PR TITLE
UltraHonk Composer (Split)

### PR DESCRIPTION
**Pull request by ledwards2225**
# Description

Introduce a complete Ultra Honk composer (following the new "split" pattern). For now, the "correctness" of the Ultra Honk Composer is confirmed by simply checking consistency with the Lagrange polynomials produced by the original UltraPlonk composer for all circuits tested in the Ultra Composer test suite. This includes all precomputed polynomials (selectors, permutation polys, table polys) as well as the wires and sorted concatenated list polynomials.

Note: This work does not include proof construction or verification but a shell of the `Prover` is included for convenience since the aforementioned polynomials are constructed in the `create_prover()` method of the constructors.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: cpp/src/barretenberg/honk/composer/composer_helper/standard_honk_composer_helper.cpp [hunk 1]
```diff
@@ -165,6 +165,7 @@ StandardProver StandardHonkComposerHelper<CircuitConstructor>::create_prover(
     compute_witness(circuit_constructor);
 
     size_t num_sumcheck_rounds(circuit_proving_key->log_circuit_size);
+    // TODO(luke): what is this manifest? Remove?
```
Comment (Rumata888) [inline cpp/src/barretenberg/honk/composer/composer_helper/standard_honk_composer_helper.cpp:168]: I guess, yes, but it is very weird that his is not raising a warning

File: cpp/src/barretenberg/honk/composer/ultra_honk_composer.hpp [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+    void create_big_add_gate(const add_quad& in, const bool use_next_gate_w_4 = false)
+    {
+        circuit_constructor.create_big_add_gate(in, use_next_gate_w_4);
+    };
+
+    // void create_big_add_gate_with_bit_extraction(const add_quad& in);
+    // void create_big_mul_gate(const mul_quad& in);
+    // void create_balanced_add_gate(const add_quad& in);
+
+    // void create_mul_gate(const mul_triple& in) override;
```
Comment (Rumata888) [inline cpp/src/barretenberg/honk/composer/ultra_honk_composer.hpp:63]: Do we not have these gates or are they not being tested?

File: cpp/src/barretenberg/honk/composer/ultra_honk_composer.hpp [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    //  * 3) Number of Rom array-associated gates
+    //  * 4) NUmber of range-list associated gates
+    //  *
+    //  *
+    //  * @param count return arument, number of existing gates
+    //  * @param rangecount return argument, extra gates due to range checks
+    //  * @param romcount return argument, extra gates due to rom reads
+    //  * @param ramcount return argument, extra gates due to ram read/writes
+    //  */
+    // void get_num_gates_split_into_components(size_t& count,
```
Comment (Rumata888) [inline cpp/src/barretenberg/honk/composer/ultra_honk_composer.hpp:134]: Don't we have these functions in the circuit constructor?

File: cpp/src/barretenberg/honk/composer/ultra_honk_composer.test.cpp [hunk 4]
```diff
@@ -0,0 +1,1015 @@
...
+ * @brief Build UltraHonkComposer
+ *
+ */
+TEST(UltraHonkComposer, test_no_lookup_proof)
+{
+    auto honk_composer = UltraHonkComposer();
+    auto plonk_composer = proof_system::plonk::UltraComposer();
+
+    size_t MM = 4;
+    for (size_t i = 0; i < MM; ++i) {
```
Comment (Rumata888) [inline cpp/src/barretenberg/honk/composer/ultra_honk_composer.test.cpp:165]: Maybe there is merit in implementing these as templated lambda functions and then just using two calls?

**Review by Rumata888 — APPROVED** [review]
There is a lot of stuff commented out (parts such as create_mul_gate) that seems important, but since there are no tests for them, I guess that's ok for now, we'll just iteratively enable them as we  implement ultrahonk.
What I'd change is the way code is duplicated in the tests, by substituting similar functionality with templated lambda functions. Other than that everything looks great)

File: cpp/src/barretenberg/honk/composer/ultra_honk_composer.hpp (hunk 2 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/composer/ultra_honk_composer.hpp:63]: I haven't investigated the extent to which they're tested elsewhere but I tried to only include functionality that's directly tested in the original composer suite. Didn't want to include too much that I wasn't also testing

File: cpp/src/barretenberg/honk/composer/ultra_honk_composer.hpp (hunk 3 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/composer/ultra_honk_composer.hpp:134]: Yeah the circuit constructor should contain virtually everything. I plan to bring these in once we have a direct test for them.

File: cpp/src/barretenberg/honk/composer/composer_helper/standard_honk_composer_helper.cpp (hunk 1 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/composer/composer_helper/standard_honk_composer_helper.cpp:168]: I'll just remove this, I think it was added by Cody by mistake

File: cpp/src/barretenberg/honk/composer/ultra_honk_composer.test.cpp (hunk 4 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/composer/ultra_honk_composer.test.cpp:165]: Totally agree but the plan is to remove the Plonk/Honk comparison aspect and convert these to traditional composer tests (full proof construction/verification) as soon as that's working.

**Pull request merged**